bigchaindb/upsert_validator/validator_election.py

```diff
@@ -2,257 +2,53 @@
-        return (current_topology == voters)
+    CREATE = OPERATION
+    ALLOWED_OPERATIONS = (OPERATION,)
 
     def validate(self, bigchain, current_transactions=[]):
-        """Validate election transaction
-        For more details refer BEP-21: https://github.com/bigchaindb/BEPs/tree/master/21
-
-        NOTE:
-        * A valid election is initiated by an existing validator.
-
-        * A valid election is one where voters are validators and votes are
-          alloacted according to the voting power of each validator node.
-
-        Args:
-            bigchain (BigchainDB): an instantiated bigchaindb.lib.BigchainDB object.
-
-        Returns:
-            ValidatorElection object
-
-        Raises:
-            ValidationError: If the election is invalid
+        """For more details refer BEP-21: https://github.com/bigchaindb/BEPs/tree/master/21
         """
-        input_conditions = []
-
-        duplicates = any(txn for txn in current_transactions if txn.id == self.id)
-        if bigchain.get_transaction(self.id) or duplicates:
-            raise DuplicateTransaction('transaction `{}` already exists'
-                                       .format(self.id))
-
-        if not self.inputs_valid(input_conditions):
-            raise InvalidSignature('Transaction signature is invalid.')
 
         current_validators = self.get_validators(bigchain)
 
-        # NOTE: Proposer should be a single node
-        if len(self.inputs) != 1 or len(self.inputs[0].owners_before) != 1:
-            raise MultipleInputsError('`tx_signers` must be a list instance of length one')
+        super(ValidatorElection, self).validate(bigchain, current_transactions=current_transactions)
 
         # NOTE: change more than 1/3 of the current power is not allowed
         if self.asset['data']['power'] >= (1/3)*sum(current_validators.values()):
             raise InvalidPowerChange('`power` change must be less than 1/3 of total power')
 
-        # NOTE: Check if the proposer is a validator.
-        [election_initiator_node_pub_key] = self.inputs[0].owners_before
-        if election_initiator_node_pub_key not in current_validators.keys():
-            raise InvalidProposer('Public key is not a part of the validator set')
-
-        # NOTE: Check if all validators have been assigned votes equal to their voting power
-        if not self.is_same_topology(current_validators, self.outputs):
-            raise UnequalValidatorSet('Validator set much be exactly same to the outputs of election')
-
         return self
 
-    @classmethod
-    def generate(cls, initiator, voters, election_data, metadata=None):
-        (inputs, outputs) = cls.validate_create(initiator, voters, election_data, metadata)
-        election = cls(cls.VALIDATOR_ELECTION, {'data': election_data}, inputs, outputs, metadata)
-        cls.validate_schema(election.to_dict(), skip_id=True)
-        return election
-
     @classmethod
     def validate_schema(cls, tx, skip_id=False):
```

Comment:
@kansi do we still need `skip_id` here?

Can we make the schema another parameter and do

```
_validate_schema(TX_SCHEMA_COMMON, tx)
_validate_schema(TX_SCHEMA_CREATE, tx)
_validate_schema(cls.TX_SCHEMA_CUSTOM, tx)
```
in `Election` and
```
cls.validate_id(tx)
_validate_schema(TX_SCHEMA_COMMON, tx)
_validate_schema(TX_SCHEMA_CREATE, tx)
_validate_schema(TX_SCHEMA_CUSTOM, tx)
```
in `Vote`?

Reply:
Done

Reply:
We can try removing it and see if the tests fail. If they do fail then we can defer removing `skip_id` to a separate PR as that is not the context of this PR.

Reply:
We're calling `validate_schema` with `skip_id=True` in the `Election.generate()` code. I'm guessing we do this because we've just computed the `tx_id` when generating the election, so no need to repeat this when validating.

In any case, it looks like the `skip` parameter serves a purpose, so let's leave it in.

Reply:
> I'm guessing we do this because we've just computed the tx_id when generating the election, so no need to repeat this when validating.

The parameter was introduced so that we can reuse the same validation function for `CREATE` and `TRANSFER` transactions. When a `CREATE` transaction is built, its ID is computed as a hash of all of its data so there is nothing to validate in the beginning. While for `TRANSFER` its ID is `asset.id` and it can be validated.

In this PR, after we generalised the election process, there are two distinct classes - `Election` and `Vote`, and each of them implements `validate_schema` differently so `skip_id` does not serve any purpose now - let's remove it.